File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] cand, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        util(cand, 0, target, path, res);
        return res;
    }
    
    public void util(int[] cand,int start, int target, List<Integer> path, List<List<Integer>> res){
        if(start >= cand.length)
            return;
        if(target == 0){
            res.add(new ArrayList<>(path));
            return;
        }
        for(int i=start;i<cand.length;i++){
            if(target-cand[i] >= 0){
                path.add(cand[i]);
                util(cand, i, target-cand[i], path, res);
                path.remove(path.size()-1);
            }
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py


class Solution:
    def combinationSum(self, can: List[int], t: int) -> List[List[int]]:
        res = []
        self.backtrack(0, [], t, res, can)
        return res
    
    def backtrack(self, curr, path, target, res, can):
        if curr == len(can) and target < 0:
            return
        if target == 0:
            res.append(path)
        for i in range(curr, len(can)):
            if can[i] <= target:
                self.backtrack(i, path + [can[i]], target - can[i], res, can)
            
        
        